x/concentrated-liquidity/math/math.go

```diff
@@ -58,19 +58,30 @@ func CalcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec
 		sqrtPriceA, sqrtPriceB = sqrtPriceB, sqrtPriceA
 	}
 	diff := sqrtPriceB.Sub(sqrtPriceA)
-	denom := sqrtPriceA.Mul(sqrtPriceB)
 	// if calculating for amountIn, we round up
-	// if calculating for amountOut, we don't round at all
+	// if calculating for amountOut, we don't round but truncate at precision end
+	// in the rounding direction.
 	// this is to prevent removing more from the pool than expected due to rounding
 	// example: we calculate 1000000.9999999 uusdc (~$1) amountIn and 2000000.999999 uosmo amountOut
 	// we would want the user to put in 1000001 uusdc rather than 1000000 uusdc to ensure we are charging enough for the amount they are removing
 	// additionally, without rounding, there exists cases where the swapState.amountSpecifiedRemaining.GT(sdk.ZeroDec()) for loop within
 	// the CalcOut/In functions never actually reach zero due to dust that would have never gotten counted towards the amount (numbers after the 10^6 place)
 	if roundUp {
+		// Note that we do MulTruncate so that the denominator is smaller as this is
+		// the case where we want to round up to favor the pool.
+		// Examples include:
+		// - calculating amountIn during swap
+		// - adding liquidity (request user to provide more tokens in in favor of the pool)
+		// The denominator is truncated to get a higher final amount.
+		denom := sqrtPriceA.MulTruncate(sqrtPriceB)
 		return liq.Mul(diff).Quo(denom).Ceil()
 	}
-	// Investigate if this should be a QuoTruncate: https://github.com/osmosis-labs/osmosis/issues/4646
-	return liq.Mul(diff).Quo(denom)
+	// These are truncated at precision end to round in favor of the pool when:
+	// - calculating amount out during swap
+	// - withdrawing liquidity
+	// The denominator is rounded up to get a smaller final amount.
+	denom := sqrtPriceA.MulRoundUp(sqrtPriceB)
+	return liq.MulTruncate(diff).QuoTruncate(denom)
```

Comment:
Since we do fixed point multiplication, `Mul` shouldn't have any rounding error right? I would have expected this line to be:
```
return liq.Mul(diff).QuoTruncate(denom)
```

Reply:
The way `sdk.Dec` is implemented is that it allows the multiplication and other math operations to go beyond the fixed point precision, then chop it and round at precision end.

So imagine we had a `Dec` multiplication `0.12 * 0. 12 = 0.0144` with the precision of 2.
After performing the operation, we would chop precision and truncate for it to become 0.01

ref: https://github.com/osmosis-labs/cosmos-sdk/blob/7a2cf2993434124b84075aafc7483b8a5ecfd9de/types/decimal.go#L303-L304

Reply:
By default, it does banker's rounding so with the above example it would still be `0.01`

However, say we have:
```
>>> 0.42 * 0.22
0.0924
```

This would become `0.1` with regular `Mut` and `0.09` with `MulTruncate`